### ISSUE TITLE
XEP-0384: OMEMO Encryption support update

### ISSUE THREAD

**Issue by Neustradamus** [mierenhoop/picomemo#1]
Dear @mierenhoop,

Good job!

It is possible to update the XEP-0384: OMEMO Encryption support to latest version?
- https://xmpp.org/extensions/xep-0384.html

XEP-0454: OMEMO Media sharing:
- https://xmpp.org/extensions/xep-0454.html

Diff Tool:
- 0.3.0 to current: http://www.aptest.com/standards/htmldiff/htmldiff.pl?oldfile=https://xmpp.org/extensions/attic/xep-0384-0.3.0.html&newfile=https://xmpp.org/extensions/xep-0384.html
- 0.8.3 to current: http://www.aptest.com/standards/htmldiff/htmldiff.pl?oldfile=https://xmpp.org/extensions/attic/xep-0384-0.8.3.html&newfile=https://xmpp.org/extensions/xep-0384.html
- 0.8.3 to 0.9.0: http://www.aptest.com/standards/htmldiff/htmldiff.pl?oldfile=https://xmpp.org/extensions/attic/xep-0384-0.8.3.html&newfile=https://xmpp.org/extensions/attic/xep-0384-0.9.0.html
- 0.8.2 to 0.8.3: http://www.aptest.com/standards/htmldiff/htmldiff.pl?oldfile=https://xmpp.org/extensions/attic/xep-0384-0.8.2.html&newfile=https://xmpp.org/extensions/attic/xep-0384-0.8.3.html
- 0.8.1 to 0.8.2: http://www.aptest.com/standards/htmldiff/htmldiff.pl?oldfile=https://xmpp.org/extensions/attic/xep-0384-0.8.1.html&newfile=https://xmpp.org/extensions/attic/xep-0384-0.8.2.html
- 0.8.0 to 0.8.1: http://www.aptest.com/standards/htmldiff/htmldiff.pl?oldfile=https://xmpp.org/extensions/attic/xep-0384-0.8.0.html&newfile=https://xmpp.org/extensions/attic/xep-0384-0.8.1.html
- 0.7.0 to 0.8.0: http://www.aptest.com/standards/htmldiff/htmldiff.pl?oldfile=https://xmpp.org/extensions/attic/xep-0384-0.7.0.html&newfile=https://xmpp.org/extensions/attic/xep-0384-0.8.0.html
- 0.6.0 to 0.7.0: http://www.aptest.com/standards/htmldiff/htmldiff.pl?oldfile=https://xmpp.org/extensions/attic/xep-0384-0.6.0.html&newfile=https://xmpp.org/extensions/attic/xep-0384-0.7.0.html
- 0.5.0 to 0.6.0: http://www.aptest.com/standards/htmldiff/htmldiff.pl?oldfile=https://xmpp.org/extensions/attic/xep-0384-0.5.0.html&newfile=https://xmpp.org/extensions/attic/xep-0384-0.6.0.html
- 0.4.0 to 0.5.0: http://www.aptest.com/standards/htmldiff/htmldiff.pl?oldfile=https://xmpp.org/extensions/attic/xep-0384-0.4.0.html&newfile=https://xmpp.org/extensions/attic/xep-0384-0.5.0.html
- 0.3.0 to 0.4.0: http://www.aptest.com/standards/htmldiff/htmldiff.pl?oldfile=https://xmpp.org/extensions/attic/xep-0384-0.3.0.html&newfile=https://xmpp.org/extensions/attic/xep-0384-0.4.0.html

Thanks in advance.

**Comment by mierenhoop** [comment]
Hi,

The choice for OMEMO 3.0 was because of the greater client support. I think it's best to upgrade to a newer version when either many clients drop 3.0 support or when the OMEMO XEP becomes stable.

On "XEP-0454: OMEMO Media sharing", that probably fits well in omemo.c, I'll look into it some more.

**Comment by Neustradamus** [comment]
@mierenhoop: The problem is if no people support latest, none will support.
So you must to support latest.

**Comment by Neustradamus** [comment]
@mierenhoop: There are already:
- Twomemo: https://github.com/Syndace/python-twomemo + https://python-twomemo.readthedocs.io/en/latest/
- Kaidan: https://www.kaidan.im/2023/05/05/kaidan-0.9.0/
- Libervia: https://libervia.org/blog/view/goffi@goffi.org/@/id/libervia-progress-note-2022-w45-MTdL
- UWPX: https://github.com/UWPX/UWPX-Client

Can you add latest XEP-0384 version?

When it will be good, the Psi/Psi+ teams (@Ri0n) can use your code to replace the libomemo-c fork which only supports "OMEMO XEP-0384 0.5.0+" (originally used in Dino):
- https://github.com/psi-im/libomemo-c

Note: You can contribute too in this code.

**Comment by mierenhoop** [comment]
I don't think this OMEMO implementation will be a good fit for the messengers you described. As stated in the README, this repo will not have support for multiple versions of OMEMO at the same time.

Maybe the newest version of OMEMO will be implemented in a new branch sometime and maybe it will even replace the master branch. Then those messengers would lose compatibility with others and it wouldn't be great for anyone.

This OMEMO implementation is also currently not as secure as libomemo-c. See it more as an experiment.

**Comment by Neustradamus** [comment]
@mierenhoop: If you can create a new branch with the latest support, it will be nice :)

**Comment by mierenhoop** [comment]
Updated to version 0.9.

I have not tested it yet with other clients, since I haven't managed to get one running that supports >0.3, but the session initiation works with python-twomemo.

**Comment by Neustradamus** [comment]
@mierenhoop: Nice, good job! :)

**Comment by Neustradamus** [comment]
@mierenhoop: Please inform me when you will create a first release build for announcement :)

Have you plan the XEP-0454 support too?

**Comment by mierenhoop** [comment]
I've added XEP-0454 support just now in a new branch. Before merging I first want to try integrating it in a client to see if it needs any changes.

**Comment by Neustradamus** [comment]
@mierenhoop: Nice!

You can look here what softwares support it:
- https://xmpp.org/extensions/#xep-0454-implementations

There is another one about OMEMO: XEP-0396: Jingle Encrypted Transports - OMEMO:
- https://xmpp.org/extensions/#xep-0396-implementations

**Comment by mierenhoop** [comment]
@Neustradamus 

I have made the first release.

**Comment by Neustradamus** [comment]
@mierenhoop: Good job! :)
Sorry for the delay of answer!

I have checked the .doap file:
- https://github.com/mierenhoop/picomemo/blob/master/project.doap

Why partial?
What is missing?

Have you checked other OMEMO XEPs?
- XEP-0396: Jingle Encrypted Transports - OMEMO: https://xmpp.org/extensions/xep-0396.html
- XEP-0454: OMEMO Media sharing: https://xmpp.org/extensions/xep-0454.html

**Comment by mierenhoop** [comment]
> Why partial? What is missing?

This library implements the cryptography for OMEMO and partly the session management, not stuff like BTBV, persistence, XML.

> * XEP-0396: Jingle Encrypted Transports - OMEMO: https://xmpp.org/extensions/xep-0396.html

I looked at this and at the time picomemo could not handle differently sized keys, so this XEP could not be implemented with the library. That issue was fixed before the 1.0.0 release.

> * XEP-0454: OMEMO Media sharing: https://xmpp.org/extensions/xep-0454.html

There's a separate branch with an implementation of that XEP. Despite the name it's not really an OMEMO specific XEP and it doesn't tie in to the OMEMO code either, so it's not a fit for this library.

**Comment by Neustradamus** [comment]
@mierenhoop: I think that you can update the support?
- https://xmpp.org/extensions/xep-0384.html

`0.9.1 (2026-04-06)`

**Comment by mierenhoop** [comment]
You're right, thanks. I modified the DOAP file accordingly.

**Comment by Neustradamus** [comment]
@mierenhoop: Perfect :)

I have done a lot of announcements about Picomemo 1.2.0.
